packages/jupyter-ai-magics/jupyter_ai_magics/utils.py

```diff
@@ -0,0 +1,47 @@
+import logging
+from typing import Dict, Optional, Tuple, Union
+from importlib_metadata import entry_points
+from jupyter_ai_magics.aliases import MODEL_ID_ALIASES
+
+from jupyter_ai_magics.providers import BaseProvider
+
+
+Logger = Union[logging.Logger, logging.LoggerAdapter]
+
+
+def load_providers(log: Optional[Logger] = None) -> Dict[str, BaseProvider]:
```

Comment:
This function contains a lot of `if log: log.X(...)` blocks. Instead, prepend this block to the function body to guarantee that `log` references a `Logger` instance:

```
if log is None:
  log = logging.getLogger()
  log.addHandler(logging.NullHandler())
```

This creates a Logger instance with a "no-op" handler that does nothing when any method is called on the Logger instance.